### DISCUSSION TITLE
From/Get coordinates for `secp256k1::PublicKey`?

### DISCUSSION THREAD

**Discussion by bczhc** [rust-bitcoin/rust-bitcoin#3648] [Q&A] [answered]
Maybe a dumb question. But I'm just wondering if there's any convenient functions to get coordinates or to convert from a (x,y) pair for a `secp256k1::PublicKey`. I'm currently playing around with Taproot tweaking, and need a PublikcKey from the sample raw coordinates listed on the tutorial website.

Now this is my way to achieve this:

```rust
    use bitcoin::secp256k1 as secp;
    pub type Point256 = [u8; 32];

    pub trait PublicKeyExt {
        fn from_coordinates(x: Point256, y: Point256) -> anyhow::Result<secp::PublicKey> {
            let mut serialized = [0_u8; 1 + 32 + 32];
            serialized[0] = 0x04; /* uncompressed */
            serialized[1..33].copy_from_slice(&x);
            serialized[33..].copy_from_slice(&y);
            let pk = bitcoin::PublicKey::from_slice(&serialized)?;
            Ok(pk.inner)
        }

        fn from_coordinate_x(x: Point256, parity: secp::Parity) -> anyhow::Result<secp::PublicKey> {
            let xkey = secp::XOnlyPublicKey::from_slice(&x)?;
            let pk = secp::PublicKey::from_x_only_public_key(xkey, parity);
            Ok(pk)
        }

        fn coordinates(&self) -> (Point256, Point256);
    }

    impl PublicKeyExt for secp::PublicKey {
        fn coordinates(&self) -> (Point256, Point256) {
            let serialized = self.serialize_uncompressed();
            (
                (&serialized[1..33]).try_into().unwrap(),
                (&serialized[33..]).try_into().unwrap(),
            )
        }
    }
```

Thanks.

**Comment by apoelstra** [comment, marked as answer]
Yeah, munually pulling them in/out of the serialization is the best way to do this.

If you just need to do it with a one-off point, you can do it by hand easier than by code -- type `04`, paste the x-coordinate, paste the y-coordinate, and you're good to go.

In general, this library is designed to support a small safe set of cryptographic operations, not to be a general-purpose math/crypto library or to support custom protocols. So working directly with coordinates is not supported. (Anything that requires working directly on the coordinates is likely to be both dangerous and inefficient, since the affine-coordinate representation of points is not a very good one for doing computations with. Converting code from a pure-math tutorial to rust-secp is an interesting application.)

**Reply by bczhc** [reply]
Alright then. Thanks for the reply!
